cpp/src/join/detail/intersection.cuh

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/device_uvector.hpp>
+
+#include <utility>
+
+#include <thrust/copy.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/remove.h>
+#include <thrust/tuple.h>
+
+#include "indexing/construction/detail/utilities.cuh"
+#include "utility/z_order.cuh"
+
+namespace cuspatial {
+namespace detail {
+
+static __device__ uint8_t const leaf_indicator = 0;
+static __device__ uint8_t const quad_indicator = 1;
+static __device__ uint8_t const none_indicator = 2;
```

Comment:
I seem to recall a compiler error when I tried this before, but I'll try it again now.

Reply:
Ah yes. Thrust doesn't like comparing pairs of `uint8_t` and `intersection_type`:
```
../../../../src/join/detail/intersection.cuh (51): error: no operator "==" matches these operands
            operand types are: const thrust::detail::uint8_t == cuspatial::detail::intersection_type
```

The alternative is casting all uses of `intersection_type` which seems uglier than necessary:
```c++
return thrust::get<0>(t) == static_cast<uint8_t>(intersection_type::leaf);
```

I also can't create the `is_quad` iterator as `column(2).begin<intersection_type>()`, because the output iterator that's passed in has no knowledge of the `intersection_type` enum, leading to this error:
```
_deps/thrust-src/thrust/detail/tuple.inl (320): error: a value of type "cuspatial::detail::intersection_type" cannot be assigned to an entity of type "uint8_t"
```

Reply:
Alternative suggestion: use a weakly-typed `enum`, or:
```suggestion
constexpr uint8_t leaf_indicator = 0;
constexpr uint8_t quad_indicator = 1;
constexpr uint8_t none_indicator = 2;
```

Reply:
nvcc is giving me this error when I try to change these to `uint8_t constexpr`, so I think our `--expt-relaxed-constexpr` isn't making it into the compile flags somehow...
```
intersection.cuh (135): error: identifier "cuspatial::detail::none_indicator" is undefined in device code
```